### PR TITLE
Core, Spark: Migrate tests that depend on ScanTestBase to JUnit5

### PR DESCRIPTION
**Pull request by chinmay-bhat**
This PR migrates tests that depend on `SparkDistributedDataScanTestBase` to JUnit5. 
`SparkDistributedDataScanTestBase` inherits from classes that ultimately inherit from `TableTestBase`. 

`TableTestBase` uses JUnit4 imports and is the base class for atleast 90+ classes across flink, core, data, and spark. So I created `ParameterizedTableTestBase`, a version of `TableTestBase` migrated to JUnit5. This is similar to #9076 

Issue: #9086

### CODE REVIEW

**Comment by chinmay-bhat** [comment]
Based on the test failures, older spark versions depend on `core` files like `DeleteFileIndexTestBase`, `ParameterizedTableTestBase`, `DataTableScanTestBase`,`ScanPlanningAndReportingTestBase`, which have been updated to JUnit5 in this PR. 

A possible solution would be to migrate the affected files in older versions to JUnit5 as well. That would be a task bigger than this PRs scope. Another solution would be to create JUnit5 versions of the above `core` files, but I'm not keen on creating unnecessary duplicates.

@nastra @Fokko 
I faced a similar problem in #9382. Any advise would be great! :)

File: core/src/test/java/org/apache/iceberg/ScanTestBase.java [hunk 1]
```diff
@@ -19,114 +19,104 @@
...
-
-  @Parameterized.Parameters(name = "formatVersion = {0}")
-  public static Object[] parameters() {
-    return new Object[] {1, 2};
-  }
-
-  public ScanTestBase(int formatVersion) {
-    super(formatVersion);
-  }
+    extends ParameterizedTableTestBase {
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/ScanTestBase.java:42]: you should be able to extend `TestBase` now

Comment (chinmay-bhat) [reply core/src/test/java/org/apache/iceberg/ScanTestBase.java]: extended with `TestBase`, still getting similar errors. 

`DeleteFileIndexTestBase`, `SparkDistributedDataScanTestBase`, `DataTableScanTestBase`, `ScanPlanningAndReportingTestBase` are all `core` files that now extend from `TestBase`, and are also used in older Spark versions. We probably need both JUnit4 and JUnit5 versions of these files to co-exist. 

I can make that a separate PR if you think it's the right approach.

File: core/src/test/java/org/apache/iceberg/TestBase.java [hunk 2]
```diff
@@ -173,7 +173,7 @@ public class TestBase {
   public TestTables.TestTable table = null;
 
   @Parameters(name = "formatVersion = {0}")
-  protected static List<Object[]> parameters() {
+  protected static List<Object> parameters() {
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestBase.java:176]: why is this change necessary?

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/SparkDistributedDataScanTestBase.java [hunk 3]
```diff
@@ -21,46 +21,41 @@
...
-      new Object[] {1, LOCAL, LOCAL},
-      new Object[] {1, LOCAL, DISTRIBUTED},
-      new Object[] {1, DISTRIBUTED, LOCAL},
-      new Object[] {1, DISTRIBUTED, DISTRIBUTED},
-      new Object[] {2, LOCAL, LOCAL},
-      new Object[] {2, LOCAL, DISTRIBUTED},
-      new Object[] {2, DISTRIBUTED, LOCAL},
-      new Object[] {2, DISTRIBUTED, DISTRIBUTED}
-    };
+  public static List<Object> parameters() {
```
Comment (nastra) [inline spark/v3.5/spark/src/test/java/org/apache/iceberg/SparkDistributedDataScanTestBase.java:38]:
```suggestion
  public static List<Object[]> parameters() {
```

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/TestSparkDistributedDataScanDeletes.java [hunk 4]
```diff
@@ -21,42 +21,39 @@
...
-  @Parameterized.Parameters(name = "dataMode = {0}, deleteMode = {1}")
-  public static Object[] parameters() {
-    return new Object[][] {
-      new Object[] {LOCAL, LOCAL},
-      new Object[] {LOCAL, DISTRIBUTED},
-      new Object[] {DISTRIBUTED, LOCAL},
-      new Object[] {DISTRIBUTED, DISTRIBUTED}
-    };
+  @Parameters(name = "formatVersion = {0}, dataMode = {1}, deleteMode = {2}")
+  public static List<Object> parameters() {
```
Comment (nastra) [inline spark/v3.5/spark/src/test/java/org/apache/iceberg/TestSparkDistributedDataScanDeletes.java:40]:
```suggestion
  public static List<Object[]> parameters() {
```

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/TestSparkDistributedDataScanReporting.java [hunk 5]
```diff
@@ -21,41 +21,38 @@
...
-  @Parameterized.Parameters(name = "dataMode = {0}, deleteMode = {1}")
-  public static Object[] parameters() {
-    return new Object[][] {
-      new Object[] {LOCAL, LOCAL},
-      new Object[] {LOCAL, DISTRIBUTED},
-      new Object[] {DISTRIBUTED, LOCAL},
-      new Object[] {DISTRIBUTED, DISTRIBUTED}
-    };
+  @Parameters(name = "formatVersion = {0}, dataMode = {1}, deleteMode = {2}")
+  public static List<Object> parameters() {
```
Comment (nastra) [inline spark/v3.5/spark/src/test/java/org/apache/iceberg/TestSparkDistributedDataScanReporting.java:39]:
```suggestion
  public static List<Object[]> parameters() {
```

File: data/src/test/java/org/apache/iceberg/io/TestGenericSortedPosDeleteWriter.java [hunk 6]
```diff
@@ -62,7 +62,7 @@ public class TestGenericSortedPosDeleteWriter extends TestBase {
   private Record gRecord;
 
   @Parameters(name = "formatVersion = {0}, fileFormat = {1}")
-  public static List<Object[]> parameters() {
+  public static List<Object> parameters() {
```
Comment (nastra) [inline data/src/test/java/org/apache/iceberg/io/TestGenericSortedPosDeleteWriter.java:65]: this should be reverted

File: core/src/test/java/org/apache/iceberg/FilterFilesTestBase.java [hunk 7]
```diff
@@ -24,67 +24,67 @@
...
 public abstract class FilterFilesTestBase<
     ScanT extends Scan<ScanT, T, G>, T extends ScanTask, G extends ScanTaskGroup<T>> {
 
-  public final int formatVersion;
-
-  public FilterFilesTestBase(int formatVersion) {
-    this.formatVersion = formatVersion;
-  }
+  @Parameter(index = 0)
+  public int formatVersion;
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/FilterFilesTestBase.java:46]: does this need to be public or can we switch it to protected?

File: core/src/test/java/org/apache/iceberg/TestLocalFilterFiles.java [hunk 8]
```diff
@@ -18,20 +18,17 @@
...
 
-  @Parameterized.Parameters(name = "formatVersion = {0}")
-  public static Object[] parameters() {
-    return new Object[] {1, 2};
-  }
-
-  public TestLocalFilterFiles(int formatVersion) {
-    super(formatVersion);
+  @Parameters(name = "formatVersion = {0}")
+  public static List<Object> parameters() {
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestLocalFilterFiles.java:30]:
```suggestion
  public static List<Object[]> parameters() {
```

File: core/src/test/java/org/apache/iceberg/TestBase.java (hunk 2 above)
Comment (chinmay-bhat) [reply core/src/test/java/org/apache/iceberg/TestBase.java:176]: Because for some tests, we only send 1 `Object` in the list, which is satisfied by `List<Object>`.
 For ex: from `DeleteFileIndexTestBase`
```
@Parameters(name = "formatVersion = {0}")
  public static List<Object> parameters() {
    return Arrays.asList(new Object[] {2});
  }
```

Having List<Object[]> does not allow for the above case as it always expects a List of Object arrays, where the Object arrays size cant be 1.

File: core/src/test/java/org/apache/iceberg/FilterFilesTestBase.java (hunk 7 above)
Comment (chinmay-bhat) [reply core/src/test/java/org/apache/iceberg/FilterFilesTestBase.java:46]: can be switched to protected. updating

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/SparkDistributedDataScanTestBase.java (hunk 3 above)
Comment (chinmay-bhat) [reply spark/v3.5/spark/src/test/java/org/apache/iceberg/SparkDistributedDataScanTestBase.java:38]: https://github.com/apache/iceberg/pull/9416#discussion_r1448313906

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/TestSparkDistributedDataScanDeletes.java (hunk 4 above)
Comment (chinmay-bhat) [reply spark/v3.5/spark/src/test/java/org/apache/iceberg/TestSparkDistributedDataScanDeletes.java:40]: https://github.com/apache/iceberg/pull/9416#discussion_r1448313906

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/TestSparkDistributedDataScanReporting.java (hunk 5 above)
Comment (chinmay-bhat) [reply spark/v3.5/spark/src/test/java/org/apache/iceberg/TestSparkDistributedDataScanReporting.java:39]: https://github.com/apache/iceberg/pull/9416#discussion_r1448313906

File: core/src/test/java/org/apache/iceberg/TestBase.java (hunk 2 above)
Comment (nastra) [reply core/src/test/java/org/apache/iceberg/TestBase.java:176]: in that case we should be able to change what's being returned here and in the other places that have been modified in this PR to `return Arrays.asList(1, 2);` right? That way we don't need to have the additional wrapping of params into `Object[]`

Comment (chinmay-bhat) [reply core/src/test/java/org/apache/iceberg/TestBase.java:176]: That would work if the inputs were only Integers. But that's not the case for all child classes.
For ex: `TestSparkDistributedDataScanDeletes` inherits from `DeleteFileIndexTestBase` has additional parameters, that need to be wrapped into `Object[]`.

```
  @Parameters(name = "formatVersion = {0}, dataMode = {1}, deleteMode = {2}")
  public static List<Object> parameters() {
    return Arrays.asList(
        new Object[] {2, LOCAL, LOCAL},
        new Object[] {2, LOCAL, DISTRIBUTED},
        new Object[] {2, DISTRIBUTED, LOCAL},
        new Object[] {2, LOCAL, DISTRIBUTED});
  }
```

Comment (nastra) [reply core/src/test/java/org/apache/iceberg/TestBase.java:176]: yes I only meant the places that are applicable and have parameters of the same type

File: core/src/test/java/org/apache/iceberg/ScanPlanningAndReportingTestBase.java [hunk 9]
```diff
@@ -34,21 +35,24 @@
...
-    extends TableTestBase {
+    extends TestBase {
 
   private final TestMetricsReporter reporter = new TestMetricsReporter();
 
-  public ScanPlanningAndReportingTestBase() {
-    super(2);
+  @Parameters(name = "formatVersion = {0}")
+  public static List<Object> parameters() {
+    return Arrays.asList(new Object[] {2});
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/ScanPlanningAndReportingTestBase.java:50]:
```suggestion
    return Arrays.asList(2);
```

File: core/src/test/java/org/apache/iceberg/TestBase.java [hunk 10]
```diff
@@ -173,7 +173,7 @@ public class TestBase {
...
     return Arrays.asList(new Object[] {1}, new Object[] {2});
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestBase.java:177]:
```suggestion
    return Arrays.asList(1, 2);
```

File: core/src/test/java/org/apache/iceberg/TestLocalFilterFiles.java [hunk 11]
```diff
@@ -18,20 +18,17 @@
...
+    return Arrays.asList(new Object[] {1}, new Object[] {2});
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestLocalFilterFiles.java:31]:
```suggestion
    return Arrays.asList(1, 2);
```

File: core/src/test/java/org/apache/iceberg/DeleteFileIndexTestBase.java [hunk 12]
```diff
@@ -35,15 +36,17 @@
...
 public abstract class DeleteFileIndexTestBase<
         ScanT extends Scan<ScanT, T, G>, T extends ScanTask, G extends ScanTaskGroup<T>>
-    extends TableTestBase {
+    extends TestBase {
 
-  public DeleteFileIndexTestBase() {
-    super(2 /* table format version */);
+  @Parameters(name = "formatVersion = {0}")
+  public static List<Object> parameters() {
+    return Arrays.asList(new Object[] {2});
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/DeleteFileIndexTestBase.java:49]:
```suggestion
    return Arrays.asList(2);
```

File: core/src/test/java/org/apache/iceberg/TestBase.java (hunk 2 above)
Comment (chinmay-bhat) [reply core/src/test/java/org/apache/iceberg/TestBase.java:176]: If we have to use `List<Object[]>` and not `List<Object>`, we can always do this:

```
  @Parameters(name = "formatVersion = {0}")
  public static List<Object[]> parameters() {
    return Arrays.asList(new Object[][] {{2}});
  }
```

Comment (chinmay-bhat) [reply core/src/test/java/org/apache/iceberg/TestBase.java]: oh just saw your previous message, got it.

Comment (chinmay-bhat) [reply core/src/test/java/org/apache/iceberg/TestBase.java:176]: removed Object wrapping when not necessary

**Comment by chinmay-bhat** [comment]
We have files that inherit from `TestBase` used in older Spark versions that are causing CI failure. Should I move them to JUnit5 like we did in #9382 ?

**Comment by nastra** [comment]
> We have files that inherit from `TestBase` used in older Spark versions that are causing CI failure. Should I move them to JUnit5 like we did in #9382 ?

Yes that would probably be the best here

**Comment by chinmay-bhat** [comment]
not sure why the one CI test is failing, all tests pass locally for me. Can you rerun the CI?

**Review by nastra — APPROVED** [review]
LGTM, thanks @chinmay-bhat for getting this done!
